rasterio/_io.pyx

```diff
@@ -828,17 +828,41 @@ cdef class DatasetReaderBase(DatasetBase):
         return sample_gen(self, xy, indexes)
 
 
+def is_zip(data):
+    """Detect zip compressed data"""
+    return data[:4].startswith(b'PK\x03\x04')
+
+
 cdef class MemoryFileBase(object):
     """Base for a BytesIO-like class backed by an in-memory file."""
 
-    def __init__(self, initial_bytes=b''):
-        """Map bytes to a file in an in-memory filesystem."""
+    def __init__(self, file_or_bytes=None):
+        """A file in an in-memory filesystem.
+
+        Parameters
+        ----------
+        file_or_bytes : file or bytes
+            A file opened in binary mode or bytes or a bytearray
+        """
         cdef VSILFILE *vsi_handle = NULL
 
-        if not isinstance(initial_bytes, (bytearray, bytes)):
-            raise TypeError("Initial bytes must be type bytearray or bytes.")
+        if file_or_bytes:
+            if hasattr(file_or_bytes, 'read'):
+                initial_bytes = file_or_bytes.read()
+            elif isinstance(file_or_bytes, (bytearray, bytes)):
+                initial_bytes = file_or_bytes
+            else:
+                raise TypeError(
+                    "Constructor argument must be a file or bytes.")
+        else:
+            initial_bytes = b''
+
+        if is_zip(initial_bytes):
```

Comment:
I'll ticket. I've read that `tgz` performance is poor: it requires uncompressing the tar file before you can read.